src/alluvium/avro.py

```diff
@@ -0,0 +1,251 @@
+        self.order = order
+        self.aliases = aliases
+
+    def generate_avro_field(self):
+        base = {
+            "type": resolve_primitive_type(self.avro_type)
+            if isinstance(self.avro_type, (AvroPrimitiveType, AvroCustomType))
+            else self.avro_type.avro_schema(),
+            "name": self.name,
+        }
+        if self.doc:
+            base["doc"] = self.doc
+        if self.default:
+            base["default"] = self.default
+        if self.order:
+            base["order"] = self.order.value
+        if self.aliases:
+            base["aliases"] = self.aliases
+        return base
+
+
+class AvroRecordSchema(AvroSchema):
+    avro_type = AvroComplexType.RECORD
+
+    def __init__(
+        self,
+        name: str,
+        fields: List[AvroRecordField],
+        namespace: str = None,
+        doc: str = None,
+        aliases: List[str] = None,
+    ):
+        if not fields:
+            raise InvalidSchemaException(
+                self.avro_type, "Cannot provide a record schema without any fields.", name=name
+            )
+        self.name = name
+        self.fields = fields
+        self.namespace = namespace
+        self.doc = doc
+        self.aliases = aliases
+
+    def avro_schema(self):
+        base = {
+            "type": self.avro_type.value,
+            "name": self.name,
+            "fields": [field.generate_avro_field() for field in self.fields],
+        }
+        if self.doc:
+            base["doc"] = self.doc
+        if self.namespace:
+            base["namespace"] = self.namespace
+        if self.aliases:
+            base["aliases"] = self.aliases
+        return base
+
+
+class AvroArraySchema(AvroSchema):
+    avro_type = AvroComplexType.ARRAY
+
+    def __init__(self, items: Union[AvroSchema, AvroPrimitiveType]):
+        self.items = items
+
+    def avro_schema(self):
+        return {
+            "type": self.avro_type.value,
+            "items": resolve_primitive_type(self.items)
+            if isinstance(self.items, (AvroPrimitiveType, AvroCustomType))
+            else self.items.avro_schema(),
+        }
+
+
+class AvroMapSchema(AvroSchema):
+    avro_type = AvroComplexType.MAP
+
+    def __init__(self, values: Union[AvroSchema, AvroPrimitiveType]):
+        self.values = values
+
+    def avro_schema(self):
+        return {
+            "type": self.avro_type.value,
+            "values": resolve_primitive_type(self.values)
+            if isinstance(self.values, (AvroPrimitiveType, AvroCustomType))
+            else self.values.avro_schema(),
+        }
+
+
+class AvroUnionSchema(AvroSchema):
+    avro_type = AvroComplexType.UNIONS
+
+    def __init__(self, elements: List[Union[AvroSchema, AvroPrimitiveType]]):
+        if not elements:
+            raise InvalidSchemaException(
+                self.avro_type, "Cannot provide a union schema without any values."
+            )
+        self.elements = elements
+
+    def avro_schema(self):
```

Comment:
You might need to include self.avro_type.value along with the avro_type of the elements. Like this:

return {
            "type": self.avro_type.value,
            "elements": <the code you have written after return currently>,
        }

Reply:
This change might be redundant due to the nature of Union type. What do you think?

Reply:
Ah so `AvroUnionSchema` was actually a special sauce I made up. In the avro spec, if you wanted to say "A value can belong to multiple types" you would say `['int', 'long', {'type': 'map'...}]`. Inspired by python typing's Union function, I decided to use this syntax instead of []. So this isn't actually a real schema but it does copy the schema abstract class because knowing that this is a union type would be useful for testing purposes.